File: C_programming/data_structs/LinkedList/LinkedList.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>

struct LinkedList {
    int size;
    struct LinkedListNode *head;
    struct LinkedListNode *tail;
};

struct LinkedListNode {
    int value;
    struct LinkedListNode *next;
};

struct LinkedListNode *CreateLinkedListNode(int value) {
    struct LinkedListNode *newNode = malloc(sizeof(struct LinkedListNode));
    newNode->value = value;
    newNode->next = NULL;
    return newNode;
}

struct LinkedList *createLinkedList() {
    struct LinkedList *newList = malloc(sizeof(struct LinkedList));
    newList->size = 0;
    newList->head = NULL;
    newList->tail = NULL;
    return newList;
}

void AddLinkedListNode(struct LinkedList *list, int value, int nodeIndex) {
    if (nodeIndex > list->size + 1) {
        printf("Index out of bounds\n");
        return;
    }
    struct LinkedListNode *newNode = CreateLinkedListNode(value);
    list->size++;
    if (list->size == 0) {
        list->head = newNode;
        list->tail = newNode;
        return;
    }
    if (nodeIndex == 1) {
        newNode->next = list->head;
        list->head = newNode;
        return;
    } 
    struct LinkedListNode *curr = list->head;
    struct LinkedListNode *prev = list->head;
    for (int i = 1; i < nodeIndex; i++) {
        prev = curr;
        curr = curr->next;
    }
    prev->next = newNode;
    newNode->next = curr;
}

void RemoveLinkedListNode(struct LinkedList *list, int nodeIndex) {
    if (list->size == 0) return;
    if (list->size == 1) {
        free(list->head);
        list->head = NULL;
        list->tail = NULL;
        list->size--;
        return;
    }
    if (nodeIndex == 1) {
        free(list->head);
        list->head = list->head->next;
        list->size--;
        return; 
    }
    struct LinkedListNode *curr = list->head;
    struct LinkedListNode *prev = list->head;
    for (int i = 1; i < nodeIndex; i++) {
        prev = curr;
        curr = curr->next;
    }
    prev->next = curr->next;
    list->size--;
    free(curr);
}

void PrintLinkedList(struct LinkedList *list, int nodeIndex) {
    struct LinkedListNode *curr = list->head;
    if (list->size == 0) {
        printf("[]\n");
        return;
    }
    if (list->size == 1) {
        printf("[%d] -> NULL\n", curr->value);
        return;
    }
    while (curr->next != NULL) {
        printf("[%d] -> ", curr->value);
        curr = curr->next;
    }
    printf("[%d] -> ", curr->value);
    printf("NULL\n");
}

int main(int argc, char *argv[]) {
    struct LinkedList *list = createLinkedList();
    char command[255];
    int nodeValue;
    int nodeIndex;
    while (true) {
        printf("Input Instruction: ");
        scanf("%s", &command);
        if (strcmp(command, "help\0") == 0) {
            printf("\tadd [value] [index] : add node to linked list\n");
            printf("\tremove [index] : remove node from linked list\n");
            printf("\tprint : print linked list\n");
        } else if (strcmp(command, "add\0") == 0) {
            scanf("%d %d", &nodeValue, &nodeIndex);
            AddLinkedListNode(list, nodeValue, nodeIndex);
        } else if (strcmp(command, "remove\0") == 0) {
            scanf("%d", &nodeIndex);
            RemoveLinkedListNode(list, nodeIndex);
        } else if (strcmp(command, "print\0") == 0) {
            PrintLinkedList(list, nodeIndex);
        } else if (strcmp(command, "exit\0") == 0) {
            break;
        } else {
            printf("Input unknown, type 'help' for list of commands.\n");
        }
    }
    return 0;
}

File: data_structs/Python_programming/queue/LinkedList.py
class LinkedList:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None

    def enqueue(self, value):
        node = LinkedListNode(value)
        if self.tail == None:
            self.head = node
            self.tail = node
            return
        self.tail.setNext(node)
        self.tail = node

    def dequeue(self):
        if self.head == None: return
        value = self.head.getValue()
        self.head = self.head.getNext()
        return value

    def printList(self):
        if (self.head == None): return
        curr = self.head
        while (curr != None):
            print(f"{curr.getValue()} ", end="")
            curr = curr.getNext()
        print()

    def reverse(self):
        if self.head == None or self.head.getNext() == None: return
        self.tail = self.head
        prev = None
        curr = self.head
        next = curr.getNext()
        while (curr != None):
            next = curr.getNext()
            curr.setNext(prev)
            prev = curr
            curr = next
        self.head = prev

class LinkedListNode:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def getNext(self):
        return self.next

    def setNext(self, next):
        self.next = next

    def getValue(self):
        return self.value

if __name__ == "__main__":
    linkedList = LinkedList()
    while(True):
        command = input("Input instruction: ")
        instruction_words = command.split()
        instruction = None
        values = []
        if len(instruction_words) == 0:
            print("No instruction provided")
        elif len(instruction_words) == 1:
            instruction = instruction_words[0]
        elif len(instruction_words) > 1:
            instruction = instruction_words[0]
            values = instruction_words[1:]
        if instruction == "enqueue":
            for value in values:
                linkedList.enqueue(value)
        elif instruction == "dequeue":
            print(str(linkedList.dequeue()))
        elif instruction == "reverse":
            linkedList.reverse()
        elif instruction == "print":
            linkedList.printList()
        elif instruction == "help":
            print("\tenqueue [value] : enqueue element to list")
            print("\tdequeue : dequeue element from list")
            print("\treverse : reverse linked list")
            print("\tprint : print list")
            print("\thelp : list valid commands")
            print("\texit : exit interface")
        elif instruction == "exit":
            break
        else:
            print("command unkown, type 'help' to list valid commands")